libraries/libstratosphere/source/diag/detail/diag_detail_log.cpp

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2018-2020 Atmosphère-NX
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms and conditions of the GNU General Public License,
+ * version 2, as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include <stratosphere.hpp>
+
+namespace ams::diag::detail {
+
+    namespace {
+
+        struct LogPrintData {
+            char *print_str_start;
+            char *print_str_end;
+            char *print_str_cur;
+            const LogMetaData &log_metadata;
+            bool is_head;
+            char buffer[40];
+        };
+
+        void PutCharacters(void *user_data, const char *chars, int chars_len) {
```

Comment:
This needs comments, to not use auto everywhere, to use const where appropriate...